### ISSUE TITLE
Need API for iterator over image array proxy

### ISSUE THREAD

**Issue by matthew-brett** [nipy/nibabel#344]
See : https://github.com/nipy/nibabel/pull/211

We may want to make an iterator called 'iteraxis' for 'dataobj' and for the Image object.

iteraxis for the image would use the cached array if present, otherwise use the dataobj implementation.

The dataobj implementation could do something like read the whole file into a memory map (uncompressing for gzip files) and slice into that, discarding the mmap after the iterator is deleted. So, there'd be an up-front memory and time cost setting up the iterator, which would be needed for each instantiation of the iterator. Of course, the user can always avoid this by loading the image data into the cache with 'get_data()'.

See subsequent discussion on the PR for more detail, particularly : https://github.com/nipy/nibabel/pull/211#issuecomment-31376534

**Comment by bcipolli** [comment]
Sounds great! This is cool stuff.

My concern is that these interfaces keep on getting lower and lower level, with lots of hidden knowledge not making its way into the code. In #211, there was quite a bit of interesting discussion about what's faster when (with file size, index, and compression all factors).

Rather than making a low-level iterator on dataobj (or perhaps in addition), would it make sense to write an iterator for the image itself that accepts an axis as an argument, and makes a best guess of which style of loading and indexing to do, based on the image dimensions and whether the file is compressed?

I feel like we could make pretty good guesses, and really help out users who haven't recently been benchmarking things.

**Comment by matthew-brett** [comment]
It's certainly tempting to do an iteraxis on the image.

I wonder if we could get away with doing something fairly crude along with some sort of tuning parameter kwarg, like `optimize='memory'` or `optimize=speed` or `optimize=memspeed` or similar.  The problem is that it's very difficult to think of some optimization method that won't do a bad job on some systems.

**Comment by bcipolli** [comment]
Is it crazy to have a `benchmark` method, that takes as input an image and an axis of iteration, and outputs the relative speed for each optimization method? Given the `iteraxis` you define above, seems very easy to write. From the user point of view, you call `benchmark`, select the method that returns the max, and go from there.

**Comment by matthew-brett** [comment]
Certainly not crazy...

**Comment by matthew-brett** [comment]
On second thoughts, maybe the `iteraxis` idea is a good one for the mailing list.   Any interest in writing the intro email?   I can do it if you'd rather.

**Comment by bcipolli** [comment]
I'll try. It's not easy!

**Comment by matthew-brett** [comment]
Yeah - it's certainly not easy - and really, it's my job, as the person who has held it up for this long.  So, if you'd like to do it, say, otherwise, I will do it in the next day or so.

**Comment by bcipolli** [comment]
This seems complex. Better to send the current designs without the upside/downside discussion? Feel free to mine this for text to send yourself!

=====

Hi all,

Efficiently indexing and iterating over an axis of an image is clearly an important use-case. Based on previous discussion, @matthew-brett opened a github issue on this. However, it's unclear what design will capture the desired use-cases, and we thought consulting the mailing list to be best.

So far, we've discussed:
* Adding an `iteraxis` iterator on the `dataobj` property of an image. `Dataobj` tries to be efficient in how it reads data from disk.
* Adding an `iteraxis` iterator on the image itself, and allowing the user to specify a higher-level kwarg (`optimize`=memory/speed/memspeed/etc), with previous knowledge about `dataobj` vs. `get_data` embedded into the code to support that.

The upside of `dataobj.iteraxis` would be control. The downside is that each user would have to do benchmarking to determine the best method to run for their system.

The upside of `image.iteraxis` would be to embed knowledge we've gained on these objects and simplify the interface for users. The downside is it's more work for us and the right choice is system-dependent. To address this, I proposed a `benchmark` method, which outputs which `optimize` method is best for the given image on the current system.

Any ideas about use-cases, or feedback on the current directions, are appreciated!

**Comment by moloney** [comment]
It would be nice to have the ability to decompress a compressed Nifti to /tmp and load from there in a transparent fashion. This would side step most of the issues about choosing the best data access method. If you are just dealing with uncompressed files then you don't really need to choose between speed/memory efficiency.

There is still the question of using mmap or not.  I would vote we avoid them if possible, or at least have an option to avoid them.

**Comment by matthew-brett** [comment]
So proposal is that `img.iteraxis(axis_no)` does something like this:

If img.dataobj is an array proxy, and array proxy refers to compressed file, uncompress file to temporary storage, make new array proxy object referring to opened temporary file -> processed_dataobj.

If img.dataobj is an array proxy, and array proxy is uncompressed filename, open file and make new array proxy backed by open file -> processed_dataobj

If img.dataobj is an array img.dataobj == processed_dataobj

Then, img.iteraxis(axis_no) is something like:

```
slicer = [slice(None)] * len(img.shape)
for row in opened_dataobj:
    slicer[axis_no] = row
    yield processed_dataobj[slicer]
```

Brendan - is that what you meant?  If so, would that work even for huge files?  That's a lot of uncompressed temporary storage?   I guess that, if the person is iterating over the slowest-changing axis in memory, then the decompression step would not be useful.  So we'd have to decide when to decompress.

**Comment by moloney** [comment]
Yeah that is basically what I had in mind.  As you point out, we can skip the decompression if it won't affect performance (due to the axis being iterated).

I can't imagine a scenario where someone would rather pay a huge performance penalty rather than use some temporary storage.  I guess we could include a keyword arg to disable the use of temp storage just in case.
